### ISSUE TITLE
Add Interface IDeconstructible

### ISSUE THREAD

**Issue by dangi12012** [dotnet/csharplang#1411]
interface IDeconstructible<T, T1>
{
           public void Deconstruct(out T name, out T1 lname)
}
//Repeat for T1 - T16 like Func<T,T1...

I really dont like how C# to expects you to have magically named methods for anything. This should be 100% clear by the use of an Interface

**Comment by HaloFour** [comment]
Deconstruction was already designed and shipped.  Proposals like this to amend it, especially to restrict it, would be dead on arrival.

Restricting deconstruction to an interface would be unnecessarily complex, annoying, force boxing and eliminate the possibility of deconstruction via extension.

**Comment by CyrusNajmabadi** [comment]
This is not a language change, but a request for a a new API (similar to ```IEnumerable<T>``` by the looks of it).

If you want this, you can file over at dotnet/corefx (though i think it's unlikely it would be taken).  Note: you're always welcome to make this interface yourself and even provide it to others through a nuget package.  I don't see any value in it being mandated in the bcl since it's totally optional and not necessary for any of these features to work.

**Comment by VisualMelon** [comment]
I don't think this sort of thing should bare an interface, because the interface conveys almost nothing (same reason you'd use a named delegate over `Action<...>`/`Func<...>`): I would never want to cast something as/abstract over the `IDeconstructable` interface, so it shouldn't exist. It's my opinion that this sort of thing should be statically determined against the concrete type (as it is), and I'd reckon that any attempt to get around the design is going to do more harm than good (not that I see much value in deconstruction without it being fallible).

None-the-less, I wholly agree with the sentiment and deplore the recognition of arbitrary method names (how can anyone report to _like_ this?)... I just wish there was a keyword attached to the magical methods so that there is _some_ indication that they aren't just a regular method (`operator` would probably do it), but I don't anticipate that being picked up for future features.

**Comment by HaloFour** [comment]
@VisualMelon 

> None-the-less, I wholly agree with the sentiment and deplore the recognition of arbitrary method names (how can anyone report to *like* this?)

I do like it.  It means that you can code to the shape rather than a nominal interface.  It enables extending other types to support the functionality.  It's pretty common in C#.  Not just enumeration, but also LINQ and `async`/`await` are pattern-driven.  You can use LINQ syntax on literally any type that exposes the methods with the signatures it expects, and you can implement just a subset of them.  You can use extension methods to make `TimeSpan` awaitable, and you can use extension methods to make `KeyValuePair<TKey, TValue>` deconstructable.  If anything the language is moving more in that direction with the shapes and extensions proposals.

>  I just wish there was a keyword attached to the magical methods so that there is *some* indication that they aren't just a regular method

Modifying the specification and the parser to deal with new types of syntax is more time consuming than using existing syntax and lowering to normal code.

Funnily enough, it seems that `IDisposable` is the one odd ball out here.

**Comment by jnm2** [comment]
@VisualMelon I think shape-based language features are cool too, very versatile.

**Comment by yaakov-h** [comment]
Wouldn't a default `SDestructible<>` shape make more sense than binding it to an interface? Once Shapes ship, of course...

**Comment by VisualMelon** [comment]
@HaloFour don't want to side-track this too much, but 'shape' (structural) features are completely opaque (as the OP points out), and the lack of explicit intent is not something to be happy about. For the record, I dislike all instances of arbitrary method lookup it in C# (`GetAwaiter`, etc.).

The point of 'just adding a modifier' is that it would (I'd presume) be pretty cheap to implement, since it is already a recognised feature, and would have a 'one-time' cost of recognising the modifier, whereafter you need only add the look up the arbitrary method names for each new magic method. The slightly higher implementation cost would (for me at least) be paid for in quality of code no end. I didn't say anything about these `operator` (just running with this keyword, I haven't really thought about it) not being extension methods: they could otherwise operate _exactly_ like a normal method, only they are now 'allowed' to be interrogated by the compiler for the purposes of `await`, and such.

As @yaakov-h says, you could keep all of the static goodness with proper `Shapes`, but still, I don't think an `SDestructible` shape would make any more sense than an `IDestructible` interface. It would be a perfectly nice technical solution.

**Comment by HaloFour** [comment]
@VisualMelon 

Adding a modifier means changing the specification to explicitly lay out the possible shapes of these methods so that the compiler can enforce them.  It means adding metadata to the output assembly so that other assemblies can recognize it, which would likely mean waiting on the BCL addition of an attribute.  LINQ would have been a serious bear to implement if each "special method" needed to be broken out and very explicitly specified and designed.

Either way, this is the normal pattern that C# follows and there seems to be little interest in adding special syntax.

If you wanted an additional visual hint as well as compiler validation of features like this you could always use an attribute and an analyzer:

```cs
[Deconstructor]
public void Deconstruct(out int x, out int y) => (x, y) = (this.X, this.Y);
```

This is what Java does with `@Override` and `@FunctionalInterface`.

**Comment by VisualMelon** [comment]
@HaloFour I would indeed have been 'OK' with having an attribute on reflection: I'm the last person to suggest new syntax unless necessary (though I'd also argue that adding a modifier would only have a technical, not a conceptual cost).

> Adding a modifier means changing the specification to explicitly lay out the possible shapes of these methods so that the compiler can enforce them.

Sorry, I don't understand. What you said sounds great, so I'm sure I'm missing something.

> LINQ would have been a serious bear to implement if each "special method" needed to be broken out and very explicitly specified and designed.

They are already explicitly defined by `IEnumerable<T>`, I'm not convinced (at present) that requiring a modified/attribute would add significantly to the design or implementation cost: if you don't wish to require a type implement `IEnumerable<T>`, but still want it to pretend to be so, then that ought to made clear: you're already paying a huge technical cost to implement the LINQ syntax at the caller end. I never use the LINQ syntax, so I'm not familiar with the use-cases for it, so I'm probably missing something. Thank you for making me aware of this particular feature, it certainly presents a problem; however, just because the compromise might be different for LINQ in isolation does mean the compromise is the right one for anything else (or not in isolation). (All of this assumes that this behaviour of the LINQ syntax is desirable)

**Comment by yaakov-h** [comment]
@VisualMelon LINQ does not require `IEnumerable<T>` in the slightest.

Try this:

```csharp
using System;

public static class Program
{
    public static void Main()
    {
        var f = new Foo();
        var r =
            from Foo a in f
            where a != null
            select a;

        Console.WriteLine(r);
    }
}

class Foo
{
    public Foo Where(Func<Foo, bool> b) => this;
    public Foo Cast<T>() => this;
}
```

**Comment by VisualMelon** [comment]
@yaakov-h aye (Halo tipped me off about that), but the point is that the methods are already well defined, so there is no conceptual cost.

**Comment by yaakov-h** [comment]
What do you mean by "well defined", and how is it any different to `void Deconstruct(out...)`?

**Comment by jnm2** [comment]
> but the point is that the methods are already well defined, so there is no conceptual cost.

What makes LINQ methods more well-defined than other shape-based methods like Deconstruct? LINQ implementation method signatures can vary wildly.

**Comment by VisualMelon** [comment]
I was merely responding to the comment HaloFour made

> LINQ would have been a serious bear to implement if each "special method" needed to be broken out and very explicitly specified and designed.

All I was trying to suggest is that there is little to no added design cost.

**Comment by VisualMelon** [comment]
@jnm2 thanks to HaloFour, nor do I now really. A few hours ago, I was living a in lovely world where LINQ was defined by `IEnumerable<T>` (never liked and never really used the dedicated syntax, but I'm still a bit surprised I wasn't aware of that; my instinct is that I don't like it, but again I'm not familiar with the use-cases). My 'hope' was that LINQ could be defined in terms of `IEnumerable<T>`, while I would contest that such is possible for deconstruction (which was the point of my original comment), but it is not so.

**Comment by jnm2** [comment]
@VisualMelon That would ruin my IAsyncEnumerable LINQ methods.

**Comment by VisualMelon** [comment]
@jnm2 sorry, you're going to have to explain that to me (sadly Rx is something I've never gone near). (Or don't, this whole discussion is my fault, but it's also kind of pointless (I've learned something at least), I just wanted to show some kinship with the OP)

**Comment by jnm2** [comment]
Not specific to Rx, but if you define LINQ in terms of IEnumerable, you can't use LINQ on anything that's not IEnumerable. Such as IAsyncEnumerable.

**Comment by VisualMelon** [comment]
Yeah, I'm not being very precise: it could be defined by/for a number of interfaces (which is fine, so long as the interfaces are meaningful in isolation), but apparently it can't, so it's a mute point.

**Comment by jnm2** [comment]
Okay. Mute != moot, though. (apologies)

**Comment by VisualMelon** [comment]
@jnm2 oh no, thank you! I've been getting that wrong for years!

**Comment by dangi12012** [comment]
There are many problems with "Magic Methods" like GetAwaiter() or Deconstruct().
The most important part is a lack of intent. If someone sees a method with 0 references he/she might just remove it.
[Deconstructible] could also be an attribute that works with the VS Compiler to omit an error if Deconstruct is not implemented.
The language is so marvellous consistently designed with only a few places where it isn't. One being magically named methods.

**Comment by HaloFour** [comment]
@dangi12012 

The IDE is perfectly capable of tracking references to those "magic" methods:

![image](https://user-images.githubusercontent.com/10102290/37881000-fa379dc6-305e-11e8-8827-9a0a68e2bc24.png)

If you find a case where the IDE doesn't track such a reference I'd report it to the [Roslyn](http://github.com/dotnet/roslyn/) repository as a bug.

**Comment by CyrusNajmabadi** [comment]
>There are many problems with "Magic Methods" like GetAwaiter() or Deconstruct().
The most important part is a lack of intent. If someone sees a method with 0 references he/she might just remove it.

1. The IDE will tell you about references to Deconstruct methods.
2. IF you remove things when there are 0 references, why would you not do the same for your special-syntax method?

> [Deconstructible] could also be an attribute that works with the VS Compiler to omit an error if Deconstruct is not implemented.

You do not need the any changes in the language for this.  You can already create your own analyzer for that purpose today.  Just create such an attribute, write an analyzer that looks for it, and do whatever checks you feel are appropriate to make you satisfied that your code is ok.

**Comment by CyrusNajmabadi** [comment]
> The language is so marvellous consistently designed with only a few places where it isn't. One being magically named methods.

Again, these are intentional and appropriate.  Otherwise, the language would need special syntax for things like "Where" "Select" "SelectMany" etc. etc.  it would need special syntax to define all the method for enumeration.  It would need special syntax to define all the methods for async/await.  

C# has *since V1* been syntax pattern based.  this is actually a good thing.  It does not lock you into any specific implementation choices on our part.  For example, your iteration types can be structs, not classes or interfaces.

If you require special syntax here, then that means the language needs to make sure that that new syntax is at least as flexible as methods are today.  That is a huge amount of language surface just to do what?  validate that a name matches?

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

Having a common analyzer for such things (maybe even as a part of Roslyn) might not be that bad of an idea.  Sort of like how Java uses special annotations like `@Override` and `@FunctionalInterface` to instruct the compiler that the annotated should meet a certain criteria otherwise it is a compilation error.  They don't drive the language feature, but they serve as guardrails for the authors of those methods.

**Comment by CyrusNajmabadi** [comment]
Roslyn created analyzers do it didn't have to be the guardian here.  The guardian is ultimately the language.  The surrounding ecosystem can drive other guardrails here if it feels like it's valuable enough.  If there is demand, it will happen.  If not, well, then why bother :D

**Comment by qril** [comment]
Deconstruction is a type cast, and it really should be done with implicit operator semantic instead of ugly magic methods. `public static implicit operator (int, int)(Point p) => (p.X, p.Y);`
There is no semantic in Deconstruct() method. It is effectively a preprocessor rule that rewrite your code.
Now we are mixing up the language semantic and preprocessor rules. This is like patching your design flaws with duct tape.

The solely purpose for the Deconstruct() magic method to exist is to be able to patch existing classes with extension methods.

**Comment by Joe4evr** [comment]
> The solely purpose for the Deconstruct() magic method to exist is to be able to patch existing classes with extension methods.

![](https://i.imgur.com/jks6Qqr.png)

Not only is that 100% inline with prior parts of the language (`GetEnumerator()` and `GetAwaiter()` can all be, in your own words, "patched onto" any type you want - you want to argue those should be implicit operators too?), but that exact use-case comes in *REALLY* handy in a great many cases, for example iterating a dictionary. Compare:
```cs
foreach (var kvp in _dict)
{
    //so much preamble just to get some sensibly named variables
    var id = kvp.Key;
    var data = kvp.Value;
    //....
}
//vs.
foreach (var (id, data) in _dict)
{
    //variables are immediately usable
}
```

**Comment by CyrusNajmabadi** [comment]
> There is no semantic in Deconstruct() method. It is effectively a preprocessor rule that rewrite your code.

This is also how async/await works as well :)

> This is like patching your design flaws with duct tape.

Or, it's just using bog standard, simple features that have been in C# from 1.0 to enable expanding on the language with far less cost.  Why come up with a new language feature to do the job, when there's an existing one that already does it for you?

**Comment by thiagomajesk** [comment]
I have to agree with @dangi12012 at least in the part that using "magical methods" just seem out of place. I don't like the current syntax either and maybe this could be reviewed before it's mainstream enough to cause compatibility problems.

Maybe we should escalate this beyond just thinking about `IDeconstructible`. 
Out of my head (without thinking too much about it) I would rather have this feature as something analogous to a Constructor/ Destructor using a similar syntax:

```c#
// Constructor
Foo(int x, int y) { }

// Deconstructor
%Foo(out int x, out int y)  { } 
// The symbol '%' is just an example taken from elixir taht uses it to represent structures
// Just like we use '~' for finalizers

// Or maybe, just convein that if every param is `out` this would be a deconstructor
Foo(out int x, out int y)  { } 
```

Those are just some example of how this syntax could be improved instead of using this ugly "magic method". Of course, I'm justing spitting suggestions here that IMHO would make this feature look a lot more csharp-like than php-like.

**Comment by thiagomajesk** [comment]
I have to agree with @dangi12012 at least in the part that using "magical methods" just seem out of place. I don't like the current syntax either and maybe this could be reviewed before it's mainstream enough to cause compatibility problems.

> The language is so marvellous consistently designed with only a few places where it isn't. One being magically named methods.

I 100% agree with this

> The IDE is perfectly capable of tracking references to those "magic" methods:

The thing is that C# is evolving beyond Visual Studio IDE (and Windows) and it would not be fair to request someone to use an IDE to get this right while others that don't will have to deal with this awkwardness.

About the "it's fine because the language already has it" (`GetAwaiter`) argument...
This proposal may be interesting to throw in here https://github.com/dotnet/corefx/issues/15469. 
And this concept seems to be used already [here](https://docs.microsoft.com/en-us/dotnet/api/microsoft.bot.builder.dialogs.iawaitable-1?view=botbuilder-3.12.2.4).

**Comment by HaloFour** [comment]
@thiagomajesk 

> I have to agree with @dangi12012 at least in the part that using "magical methods" just seem out of place.

> I 100% agree with this

Except that his assertion that "magic methods" are the exception is incorrect.  As stated several times in this thread already C# almost universally uses "magic methods" for language features and always has.

**Comment by thiagomajesk** [comment]
@HaloFour I understand that. But please consider that maybe (just maybe) defending a "wrong" approach because it was already done in the past is not so good of an argument to pushing things forward.

We'd have to agree that C# being a statically typed language it's imperative that we use those kinda of contracts to specify APIs (Well, this is not what interfaces are all about?). 
I mean, I don't mind the usage of "magic methods" or whatever witchcraft is necessary to get the internals working. Programming in a static language requires us to rely upon well-defined contracts.

**PS.:** @HaloFour If you check out the links at the end of my previous comment, you will see that this is a reality and we are walking towards that path. 

The impression it gives me is that this was not well discussed as it should have been and they've rushed the decision. Perhaps this is a good approach to solve past problems with "magic methods" as well.

**Comment by CyrusNajmabadi** [comment]
> But please consider that maybe (just maybe) defending a "wrong" approach because it was already done in the past is not so good of an argument to pushing things forward.

I think the issue is tht people don't agree it's a "Wrong" approach.  This approach has worked well for 15 years and hasn't actually resulted in any problems.  So i think the onus is on you to actually make a *convincing* argument that there's a problem here.

> at least in the part that using "magical methods" just seem out of place. 

This has been how the language has worked since literally v1 i think (and certainly v2 onwards).  It sseems weird to claim that these are "out of place" given that it's been hte normal for 15 years :)

**Comment by thiagomajesk** [comment]
@CyrusNajmabadi If consistency is not a good argument, I don't know what else would be.  
One of the things that C# does really well is that you know what to expect most of the times. 
(This of course will no longer be true if the team will just keep pushing this kind of things).

> This approach has worked well for 15 years and hasn't actually resulted in any problems

This is exactly the kind of argument I'm trying to deconstruct here (no pun intended). Just because this worked in the past and maybe people let those ugly bits slide, it doesn't mean that overall all these decisions were the best decisions (design-wise). 

What I'm generally trying to say is that the team designing the language should think a lot more about the users towards consistency and programming-happiness than practical decisions. 
We have a nice case for this and probably anyone from the PHP community can lecture about how this design decisions worked for them.

> So i think the onus is on you to actually make a convincing argument that there's a problem here.

I don't think that way. Maybe you could change your mind around the fact that are a lot of reasons to "fix" something even if it's not broken. I'm not agreeing to dangi12012 with the intent to prove that there's a "problem". I do share the vision that another path would be more concise regarding the way this is feature is presented.

**Comment by CyrusNajmabadi** [comment]
I find it weird that the term "magic method" is bandied around.  The point of the design of these features is that the methods are *not* 'magic'.  That's the core virtue of them.  They're just plain, ordinary methods.  This is great.  It means all the stuff you know about methods applies to them.  It means you can use them like normal methods.  It means you can find them *the exact same way* you find normal methods.  It means things like "extension methods for linq" "just work".  Similarly, it's why 'async/await' can be so flexible. 

Using normal methods for this sort of thing has been part of C#'s DNA *since hte start*. If this was v1/v2, there could be more weight thrown into the "this isn't a good idea" and "this will end up being a problem" argument.  But we're already 15 years past that point.  The reality is that not only has this not been a problem, it's actually proven to be an fantastic way to solve things.

That's why the LDM keeps going back to solving problems using this mechanism.  Because the mechanism has a *proven* track record of being a great way to solve problems.   Suggesting that the language and LDM turn its back on what has worked out so well for so long is nonsensical.

**Comment by thiagomajesk** [comment]
@CyrusNajmabadi Well, you apparently know a lot more of the reason behind those decisions than us.
And since this back and forth arguments won't change anything at this point, basically. 
Would you mind shifting a little bit of the conversation towards what limitations **you** think/ know would happen if we forced an interface on this? I feel that at least when this Issue is closed, we will have a better use-case history for this types of discussions.

**PS.:** I saw a lot of discussion about pattern-based and interface-based features, and for what I see you are basically putting the former above the latter. So, I'd like to you to elaborate on that a little if you may. (At Whom it may interest: https://github.com/dotnet/csharplang/issues/1467).

**Comment by CyrusNajmabadi** [comment]
> PS.: I saw a lot of discussion about pattern-based and interface-based features, and for what I see you are basically putting the former above the latter. So, I'd like to you to elaborate on that a little if you may. (At Whom it may interest: #1467).

Basically: patterns are more flexible, allow for faster code, and do not control yoru actual inheritance hierarchy.  They also decouple the type from the operations on it, which is great for allowing *others* to use types that they did not author in ways they feel make sense.

i.e. if i make a Point type, but i didn't think to make it deconstructible, someone else can do that themselves.    This is not a hypothetical virtue.  In roslyn itself we have introduced this for types that we would like to deconstruct but which we were not the author of.

**Comment by CyrusNajmabadi** [comment]
> Would you mind shifting a little bit of the conversation towards what limitations you think/ know would happen if we forced an interface on this scenarios?

I mean... just think about linq, and had it been an interface, and not just 'normal methods'.  You literally would not have been able to use 'linq' on IEnumerables, which would have made the feature (IMO), utterly useless.

**Comment by thiagomajesk** [comment]
Ok, maybe my answers here will be a little bit naive, but just to put these things to rest... 

> As mentioned above if this requires an interface you cannot add this functionality to some other arbitrary type via extension methods. All of LINQ depends on that.

Why would you want that instead of just implementing the interface?
If you don't own the type, you should not be able to change its "shape". That's why we use extension methods right!? I mean: Instead of breaking the default behavior we only add more functionality.
The way I see, you have one alternative: Inheritance. See Java prior to version 7 where the only way to extend behaviour was through inheritance (now they have default interfaces implementations for that).
But still, if you'd want to change the shape of the class you probably should create another one - a different one, with common, but different behavior.

> For another, I think it's limiting when you put something in that adds no value. It's unclear to me what value the actual IDeconstructible interface would have. What do you actually need it for?

Well, as you know an interface is an explicit contract to something; and this should be the only way to expect behavior from our classes. We have: `IEquatable<T>` for _Equals()_;  `IEnumerable` for _GetEnumerator()_ and `IDiposable` for _Dispose()_. This way you'll always know what to expect. 
Everything here is explicit because **you know** there is a contract/ behavior you're signing in to. 
The only difference from the current usage of `Destructor` is that you don't have the static guarantee of the contract and now you have to rely on the IDE, analysers, etc . So this is no different than a dynamic language anymore, you are "magically" adding behaviour.

> Interfaces are important when you want to abstract away and compose operations over any impl. No one has given an actual use case where you would do that with deconstructible objects. What is the actual (not hypothetical) use case here?

Explicitly enforce a behavior? If you are not using interfaces for that what is taking us from making everything in C# just like a dynamic language with this methods that express behavior but are not tied to an interface?

> I would like to see actual code that people want to write (either here or in a gist) that demonstrates the value of this interface.

Maybe this is not the best example, but for starters we could deconstruct every object on assembly that implements the `IDestructible` interface. I can search for that interface and this assures me that I can expect the behaviour of those objects, and I have a compile time guarantee of that.

> i.e. if i make a Point type, but i didn't think to make it deconstructible, someone else can do that themselves. This is not a hypothetical virtue. In roslyn itself we have introduced this for types that we would like to deconstruct but which we were not the author of.

See my previous answer to that. If you don't own the type why do you want to change the way it is? 
Just create another one or use inheritance. Being able to change a class like that - not extending, but actually changing its default behaviour like this is a problem (see Ruby open classes; sure make things easier but it can introduce unespected behavior).

> I mean... just think about linq, and had it been an interface, and not just 'normal pattern-based methods'. You literally would not have been able to use Linq on IEnumerables, which would have made the feature (IMO), utterly useless.

How so? `System.Linq` is basically a bunch of extensions on top of  [`IEnumerable`](https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Aggregate.cs) and you sure can add extensions to interfaces. What do you mean by that?

**Comment by CyrusNajmabadi** [comment]
> Why would you want that instead of just implementing the interface?

Many reasons:

1. That implies you *can* even implement the interface.  What if this is code in binary form that shipped 5 years ago and you can't update?  What if hte owner doesn't want to implement the interface?
2. This implies you *should* implement hte interface.  A huge virtue of Linq is that it didn't require everyone under the sun to implement it.  Instead, it could be implemented *for free* for most libraries.  You only implemented the pattern yourself if you wanted custom domain logic.

Basically, your question boils down to: what's wrong with doing a bunch of unnecessary work when we have a solutoin that requires none of that?  

> If you don't own the type, you should not be able to change its "shape"

Then linq would have been useless.  After all, most people do not 'own' System.IEnumerable.  And you can't even change System.IEnumerable as that would break the thousands of implementations of that interface *everywhere*.

> you should not be able to change its "shape"

What an awful state to be in.  Should i not be able to decosntruct a point if someone else defined it?  Should i not be able to run a linq query over an object if i can appropriately map it to teh linq pattern?

Why do i need to own a type in order to be able to use these language patterns on them?

**Comment by CyrusNajmabadi** [comment]
> We have: IEquatable<T> for Equals(); IEnumerable for GetEnumerator() 

That is not true.  

The language has not required IEnumerable at all for GetEnumerator.  Indeed, there are many times that people do not use IEnumerable for GetEnumerator *precisely* because they don't to potentially box through the interface.

**Comment by CyrusNajmabadi** [comment]
> Explicitly enforce a behavior? If you are not using interfaces for that what is taking us from making everything in C# just like a dynamic language with these methods that express behavior but are not tied to an interface?

The *majority* of methods in the BCL are not tied to an interface.  So what?  :)    That in no way has made the BCL particularly 'dynamic'.  

An interface has never been necessary to express behavior.  Indeed, it can't... since it doesn't have code in it.  An interface simply states what operations you expose, but makes no claims beyond that.

**Comment by CyrusNajmabadi** [comment]
> Maybe this is not the best example, but for starters, we could deconstruct every object on the assembly that implements the IDestructible interface.

For what purpose?  This is like me saying "hte IFrob interface is useful so that i know i can frob() objects that implement it".  Ok... that's a truism.  But how is that *useful*.  What problems are you solving with this interface?

For example, i've mentioned cases where i want to operate agnostically on arbitrary streams of data (that's literally what Linq is). In that case having an interface that can be referred to here is important.  I'm looking for the use case where you *need* to know the type is deconstructible and what you would then do in response to that.

**Comment by CyrusNajmabadi** [comment]
>> I mean... just think about linq, and had it been an interface and not just 'normal pattern-based methods'. You literally would not have been able to use Linq on IEnumerables, which would have made the feature (IMO), utterly useless.
> How so? System.Linq is basically a bunch of extensions on top of IEnumerable and you sure can add extensions to interfaces. What do you mean by that?

I mean just what i said.  Go ahead and try.  Make linq work if you need to do it with some new interface.  For example, say we start with ```IEnumerable<T>```.  Now you add:

```c#
interface ILinqEnumerable<T> : IEnumerable<T> {
     ILinqEnumerable<T> Where(Func<T> pred);
}
```

Ok... great... now you have this interface.  Except no one can use it.  Because nothing in the world implements this interface.  So you can't actually '.Where' any existing collections out there.  Because they don't actually impelment this interface.

That's why linq is not done with interfaces, but is instead done with *patterns*.  It allows the entire system to work without mandating anything about the implementation of the actual type you want to operate on.

**Comment by CyrusNajmabadi** [comment]
> See my previous answer to that. If you don't own the type why do you want to change the way it is?

The second part of your sentence has nothing to do with the first.  It's like saying "if you don't own the type, why would you put it in a collection?"   Because that's what i want to do with it :)    I want to be able to pattern match/deconstruct a type because... well.. that's useful to me.  It allows me to stuff like say:

```c#
switch (point) {
    case (0, 0): // handle origin case; ...
    case (var x, 0): // handle anything on the x axis ...
    case (0, var y): // handle anything on the y axis ...
    case (var x, var y): 
}
```

Why should i not be able to do that just because the original person who wrote point didn't happen to make it deconstructible?

> Just create another one 

Great.  Now i need to do *more* work.  It's extra code and it's extra cost.  Two things i don't like :)

> or use inheritance. 

1. only possible if this is a non-sealed class.  
2. This then means i am screwed if someone else passes me points.  Say they give me a list of Point instances.  Great, now i need to allocate N subtypes.  Why would i want to do this?

> Being able to change a class like that - not extending, but actually changing its default behavior

I changed nothing about it's behavior.

>  like this is a problem (see Ruby open classes; sure make things easier but it can introduce unexpected behavior).

What problem?  No behavior of the type changed.  

Do you literally have a problem with someone writing the following function:

```c#
if (TryGetName(customer, out var name)) { }
// ...
bool TryGetName(Customer c, out string name) { ... }
```

?

Did i change anything about the behavior of customer by having this helper method?  Not at all.  So what's the problem with it?  And if there's no problem with it, why is it a problem to have an equivalent helper called "Deconstruct" that i use *for the exact same purpose*?

**Comment by HaloFour** [comment]
@thiagomajesk 

> If you want to create a custom class that can enumerate you just have to implement IEnumerable because that's the behavior you want to expose.

It's not implementing `IEnumerable` that gets you support in LINQ.  That interface provides absolutely none of the necessary members.  It's the fact that LINQ is built on patterns and that those patterns can be satisfied via extension methods that makes LINQ possible at all.  And no, you don't need to implement `IEnumerable`.  You need to provide a way to resolve those members.  This is what makes so powerful and so applicable to such a wide variety of data types.

> You are missing the point here. I'm not saying that you should use interfaces for this kinda of cases.
You would naturally use extensions for... Well, extending the interface with new functionality without breaking who already sings to that contract, this is how it is implemented today.

Congratulations, you've just described shapes.

Either way it doesn't really matter.  C# has been designed following this philosophy for the past ~18 years and it has worked out exceptionally well.  It's most definitely not going to change.  If anything it will continue to expand by allowing developers to define their own shapes/concepts and apply it to their own APIs.

**Comment by thiagomajesk** [comment]
@HaloFour 
> It's not implementing IEnumerable that gets you support in LINQ. That interface provides absolutely none of the necessary members.

I don't know how did you get this from my answers. I'm definitely **not** saying that you should implement `IEnumerable` to get linq (like `Where`, `Select`, `Aggregate` were the methods on `IEnumerable`). I know that the interface `IEnumerable` provides only the contract for **enumerating** things. My current understanding is that linq extends `IEnumerable` (`IQueryable`, `IList`) with extensions methods.

**Comment by HaloFour** [comment]
@thiagomajesk 

> My current understanding is that linq extends `IEnumerable` (`IQueryable`, `IList`) with extensions methods.

LINQ doesn't extend anything.  It works by looking for the shape of specific methods.  It so happens that those extension methods satisfy that shape, but you are certainly not required to use those extension methods in order to use LINQ.  You don't even have to be targeting a version of the .NET framework where those extension methods exist at all.

**Comment by thiagomajesk** [comment]
@HaloFour 

> LINQ doesn't extend anything. It works by looking for the shape of specific methods

How so? For me, those are just extensions mutating data: https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Aggregate.cs
I feel that maybe you guys are talking about that in such a philosophical way that does not make practical sense of what I'm seeing implemented.

The whole discussion I think goes towards the awkwardness of having this special case for destructuring. 
The methods on linq are basically just common methods (following the previous understanding) that process and return stuff. The case for Destruct is that it does not return anything, is just there to magically satisfy a syntax sugar (the impression it gives at least).

**Comment by HaloFour** [comment]
@thiagomajesk 

> How so? For me, those are just extensions mutating data: 

When you have the following code:

```cs
var query = from value in e
    where value % 2 == 0
    select value;
```

The compiler attempts to resolve it to the following:

```cs
var query = e.Where(value => value % 2 == 0);
```

The compiler doesn't care in the slightest where `Where` comes from.  It may resolve to an instance method.  It may resolve to an extension method.  All the compiler cares about is that it can be resolved.

This shape and translation is critically important because it is what enables LINQ query providers to work as the above code and invocation to `Where` changes completely depending on whether the argument is a delegate or an expression.  In the latter case the compiler will emit an expression tree which can then be parsed by a queryable provider in order to translate the logic into some other form, like SQL.

> I feel that maybe you guys are talking about that in such a philosophical way that does not make practical sense of what I'm seeing implemented.

The details are important.  Those extension methods only matter to the extent that they satisfy the shape expected by the compiler.  They happen to be the default implementation for `IEnumerable<T>` that ships with .NET Framework 3.5 and greater.  But you can still use LINQ with .NET 2.0, you just have to satisfy the shape through some other implementation.  The vast majority of language features in C# work this way.

> The case for Destruct is that it does not return anything

Um, the entire point of deconstruction is to return stuff.  You are extracting the values from some composite type.

**Comment by thiagomajesk** [comment]
@HaloFour 

> The compiler doesn't care in the slightest where Where comes from. It may resolve to an instance method. It may resolve to an extension method. All the compiler cares about is that it can be resolved.

Well, thank you. Now, this makes a lot more sense because you are comparing linq (using the keywords) with the fluent interface (that are just extension methods). Knowing that this will be resolved for any implementation is a game-changing information.

> Um, the entire point of deconstruction is to return stuff. You are extracting the values from some composite type.

Maybe I was not sufficiently clear here. Now I can agree on how this works internally and indeed makes sense how things are resolved. Although you'll have to agree with me that while using LINQ trough keywords, it just seems like you're using a keyworded-shortcut for the `Where` method. And in the case of the Destructor, the syntax does not resemble how a method is commonly used (since there's no return). Would be different for example if the Destruct method was something like this: 
`(string, string) Destruct ()` -> `var (a, b) = destruct x` (I'm not proposing a new syntax, just making the correlation of usage between both cases).

**Comment by CyrusNajmabadi** [comment]
> Did you even consider the intentions behind it

Yes, i considered it.  And i addressed the deficiencies i felt were there with the points i made.  Do you take issue with the actual points and argument i've put forth?

> This is not how things are supposed to work?

This is how the world works (especially .net) :)   Linq was *explicitly* designed so you could work with the millions of lines of C# that had already shipped without all those millions of lines needing to be updated to support Linq.  That's one of the virtues and precisely one of the reasons a non-interface-based approach was taken for it.

> Ok, if the person who has created the Point class don't want to provide the behavior for deconstruction?

Can you give an example of what you mean by that?  They already exposed everything i needed for deconstruction.  Why can't i deconstruct it just because they didn't provide a deconstructor themselves?

Whart is the "the behavior for deconstruction"?

> The contract is there if the language is not using it internally is another matter

The contract is there so that people can compose over streams of data.  Something which is valuable and has clear uses cases for (*like linq*).  The language itself does not care about it.  The same is not true for deconstructability.  No use case has been made to be able to compose over deconstructible things.  Without a use case *what value does hte interface provide*?

> For instance, why should you be able to use Dispose on a type if the person who wrote that don't want it to be disposable?

There is no connection.  In one case you are saying "you can only use hte public surface area of a type ".  I'm saying: i should be able to deconstruct a type on my own using the public surface area they provide.".  These are compatible positions.  

> Maybe for some reason the person who created the class didn't want to expose that behavior, 

You keep saying "expose that behavior".  I do not know what that means.  What behavior?  Why do i need 'behavior' in order to deconstruct anything?

> or just wanted the class to be used in a different way, it can happen.

Can you give an example of this wrt deconstruction?  How would a person make a type, provide ways for me to get all the data from it, but then not expect or want me to then actually deconstruct that type into the data?

Let me give an example.  Say tehre is a "Point" type.  Do you have an issue with me writing:

```c#
var x = point.X;
var y = point.Y;
```

Do you have any issue with that?  I'm just using the surface area provided by the type.  Ok, assuming you have no problem with that, what is the problem with me then writing:

```c#
var (x, y) = point;
```

?

Why is the former ok, but the latter is not.

> A good example could be if you create a Point type where you want to override the + operator. Then, you say that this Point class needs to always sum +1 along with the result.

There is no "overriding".  The language always defers to the instance first if the instance provides the capability.  This is how linq works.  It's why .Where on a Queryable calls the right method instead of calling the extension method that works on IEnumerable.

**Comment by CyrusNajmabadi** [comment]
>  My current understanding is that linq extends IEnumerable (IQueryable, IList) with extensions methods.

As HaloFour mentioned, this is not the case.  There is no Linq interface.  And the compiler has now knowledge or care of *any* interfaces when dealing with Linq.  Linq is literally the compiler just saying "can i find a Where/Selection/GroupBy/GroupJoin/etc. method that i can pass the right arguments to".

That's all :)

When the compiler sees ```from a in b where a > 10 select a``` it just translates that to ```b.Where(a => a >  10);```.  That's it.  There need be no interfaces involved at all.  This is *exactly* the same with deconstruction.  When the compiler sees:  ```var (x, y) = point```, it simply translates that to ```point.Deconstruct(out var x, out var y)```.  

They are the same mechanism.  Just that the syntatic forms for linq looks a certain way in C#, and the syntactic forms for deconstruction look another.  But the same 'ordinary' *non-magical* method based approach to resolution and compilation apply.

**Comment by CyrusNajmabadi** [comment]
> The case for Destruct is that it does not return anything,

It definitely returns things.  Otherwise it would serve no purpose :)

> is just there to magically satisfy a syntax sugar (the impression it gives at least).

Yes.  Just like Linq.  :) You can't do ```from a in b where a > 10 ...``` unless there is a .Where method somewhere.  Similarly, you can't deconstruct (except for tuples i think), unless there is a .Deconstruct method somewhere.

**Comment by CyrusNajmabadi** [comment]
> Knowing that this will be resolved for any implementation is a game-changing information.

Right :)   And, importantly, this is how the language works *most* of the time.  There are some unfortunate exceptions, but theses will likely be addressed in time.  There's just too much value in doing the non-interface route.  For example, this is a critical part of the system that is used for performance.  If, for example, the compiler only called through IEnumerable.GetEnumerator, it would have devastating consequences for performance.  

Similarly, Linq literally would not work if interfaces were required to describe the linq pattern.

**Comment by CyrusNajmabadi** [comment]
> Although you'll have to agree with me that while using LINQ trough keywords, it just seems like you're using a keyworded-shortcut for the Where extension method.
But in the case of the Destructor, the syntax does not resemble the implementation (since there's no return). Would be different for example if the Destruct method was something like this:
(string, string) Destruct () and then var (a, b) = destruct x.
(I'm not proposing a new syntax, just making the correlation of usage between both cases).

Note: there is ample precedence for this.  When you write ```foreach (var x in y)```, this is translated to ```var __enumerator = y.GetEnumerator()```.  There is no 'foreach' in the translation.  There is no 'enumerator' in the original source.

>  for example if the Destruct method was something like this: (string, string) Destruct () 

This was considered, but is unpalatable for other reasons.  First of all, the language does not have a concept of overloading on return type.  So if you want your type to be deconstructible in multiple ways (for example, maybe a Point3 that allows deconstruction to '(x,y)' as well as '(x, y, z)'), then you cannot provide both those methods.  The current form was picked precisely because it enables all the scenarios people care about, and fits entirely into the incredible wealth of work that has gone in in the last 15 years into C#'s understanding of methods, overloads, arguments, etc.

``` and then ```var (a, b) = destruct x```.

That's certainly a reasonable position to take.  But the LDM simply did not feel like this was appropriate or worthwhile.  A large amount of thougth goes into: "how will we feel about this 10 years from now?"   i.e. will people just be so annoyed that htey have to write :  ```Foo(deconstruct x, deconstruct y, deconstruct z)``` over and over again.  We have precedence for making these sorts of mistakes (i.e. anonymous delegates) and we spend a lot of time working toward making clear and concise syntax that we feel will still be relevant and useful a decade from now.

Note: there is little precedence between needing a strong connection between method-name and the language syntax.  For example ```using(x)``` (this calls ```Dispose```).  Also ```foreach (...)``` this calls ```GetEnumerator```.  ```fixed``` statements do not call anything with ```Fixed``` in the name.  'Collection initializers' use ```{ {a, b}, {c, d} }``` but translate that into ```Add(a, b); Add(c, d)```.  ```lock``` calls ```Monitor.Enter```, etc. etc.

The LDM picks the syntax they think makes sense for the language and works with the framework to define the pattern that makes sense for the ecosystem of .net languages out there.

**Comment by CyrusNajmabadi** [comment]
> I feel that maybe you guys are talking about that in such a philosophical way that does not make practical sense of what I'm seeing implemented.

I disagree.  This is exactly *not philosophical*.  I'm not pushing this approach because of some vague hand-wavy feel-good reason for why this is good.  I'm listing out practical and important reasons why we do things this way.  Form the performance, to the extensibility, to the decoupling, etc. etc.  I've also provided concrete reasons why thsi approach was necessary, otherwise existing *massively important* features (like Linq) simply would not work, or would have significant drawbacks associated with them (like all the perf problems inherent with interfaces).

**Comment by CyrusNajmabadi** [comment]
> And you sure can add extensions to interfaces. What do you mean by that?

Note: extensions don't add anything to interfaces.  :)   This detail matters, especially in the context of: 

"Ok, if the person who has created the Point class don't want to provide the behavior for deconstruction?"

Now imagine i came to you and said "waht if the person who created the Frob class didn't want to provide the behavior for Linq"?   Turns out, what they want doesn't matter.  As long as you provide the linq *pattern* (not "the linq interface"), then you can use linq on those Frobs.  

Indeed, that's how Linq works int he first place.  No one asked IEnumerable "do you want to be usable in Linq".  No one went to all the types that implement IEnumerable and asked them.  Instead, we just added hte linq pattern and it all "lit up".  

Now, the same holds for deconstruction.  Just as 'linq' lit up on IEnumerable without IEnumerable (or its implementations) doing *anything*, deconstruction can light up on any type without that type explicitly doing anything itself.  If you're ok with it for linq, the same reasoning holds for deconstruction.

**Comment by CyrusNajmabadi** [comment]
> PS.: Now I makes sense to me, but I'm still finding it weird for different reasons (like I have explained)

It helps to understand how widespread this is and how common we've used this approach.  it's not like deconstruction is strange in this regard.  It is precisely in line with so many other language features (which i guess you didn't even realize worked this way :)).  That's why it seems strange when it's really just like so much other stuff in the language :)

**Comment by yaakov-h** [comment]
If I may...

Using patterns/shapes instead of interfaces also adds one very valuable aspect to any new pattern-based language feature:

It means that you can start using that language feature just by upgrading your compiler.

You don't need to wait for or upgrade all your projects to target a new version of .NET Framework (or now, .NET Core, .NET Standard, Mono, Xamarin, etc.).

You don't need all your customers/users to upgrade to a new runtime that includes that interface.

All you need to do is write methods that match the pattern - or import them from NuGet - and then start using the new language feature.

Requiring a new runtime (a la `async`/`await` in .NET 4.5) is typically a major headache for developers and would massively delay adoption of new features, which would also delay feedback and thus iterations of the features to improve upon them.

**Comment by thiagomajesk** [comment]
@CyrusNajmabadi 

> >Did you even consider the intentions behind it
> 
> Yes, i considered it. And i addressed the deficiencies i felt were there with the points i made. Do you take issue with the actual points and argument i've put forth?

I'm sorry, but you are not making this clear anywhere here. 
The really helpful explanation to put this conversation on context happened after @HaloFour jumped in the conversion to make **your** point clearer about the internal usage of patterns.
You may think you have addressed many issues with my arguments here, but you certainly missed that one key information was not present on my side and you didn't care to address that specific point. 
You have completely missed the signals I was giving - you could just have addressed that directly like @HaloFour did. Also: You didn't need to address any more of my previous comment like you did if you're considering that the main context was explained already; this just proves that you are here to **win** the argument instead of providing information to a helpfull debate.

@yaakov-h Thanks for your contribution. 
I think that the reason behind the decisions was made, are a bit clear now.

I'm getting that the deconstruction functionality was not meant to be a "closed opt-in" behavior for a type. Instead, it's extrapolating to be extensible by whom is _using the type_. In that case, using an interface to "limit" how the shape is formed does not apply.

I'm just curious though: Besides flexibility (because we can transform data inside the `Deconstruct` method) why our deconstruction is not simply like Python's, ES6's or Elixir where you're a matching against the "public" members of the type _automatically_?

**Comment by HaloFour** [comment]
@thiagomajesk 

> I'm just curious though: Besides flexibility (because we can transform data inside the `Deconstruct` method) why our deconstruction is not simply like Python's, ES6's or Elixir where you're a matching against the "public" members of the type *automatically*?

C# is getting this also in C# 8.0, as a part of recursive patterns.  But since members don't have a specific order you have to specify the member name:

```cs
if (person is Student { Gpa: 4.0, FirstName: var firstName }) {
    Console.WriteLine($"{firstName} has a perfect GPA!");
}
```

Deconstruction is intended to be positional so that it works well with tuples (where the names don't really matter) and with data classes as an inverse of construction where there are few members.

```cs
Point pt = new Point(3, 4);
var (x, y) = pt;

Dictionary<int, string> dict = ...;
foreach (var (key, value) in dict) {
}
```

At some point C# will likely also get discriminated unions or active patterns where a named conditional deconstructor will be able to convert another type and extract values from it:

```cs
Cartesian cartesian = new Cartesian(x, y);
if (cartesian is Polar(var r, var theta)) {   
}

Option<string> opt = ...;
if (opt is Some(var value)) {
}
```

A lot of this is building out the larger epic of pattern matching in the C# language.

**Comment by jnm2** [comment]
@thiagomajesk 

> You may think you have addressed many issues with my arguments here, but you certainly missed that one key information was not present on my side and you didn't care to address that specific point.
You have completely missed the signals I was giving

From an observer's point of view, I did see signals from you but not those signals. It looked like both Cyrus and HaloFour had each tried to tell you multiple times before the time when HaloFour's attempt seemed to click for you. I appreciate that it feels uncomfortable to have been making an argument while missing critical information. Think of it from the other perspective: it's really hard to guess what you already know and don't know.
We all make mistakes and should assume we're making at least as many as we perceive others to be making. Let's give each other the benefit of the doubt!

**Comment by jnm2** [comment]
I have been aware of the underlying mechanism for a few years, so yes, I'd say that biases me! 😃

> And if you care to consider

Just want to point out that “if you care to” is an argumentative tone which drowns out the signals you wanted to communicate, and there are examples of this in your responses to the others.

**Comment by CyrusNajmabadi** [comment]
> Then, there's a strong responsibility to make sure the context of the argument always is clear for both parts before even elaborating it.

THat's why i spent a lot of time giving many different examples to try to help make things click for you.  However, even when i responded, you lashed at at me.  I never treated you in a disrespectful fashion.  I put a lot of time and effort into trying to find way to help explain the matter.  I brought up many different facets that went into the decisions here. I explained hte pros and cons.  I detailed the history.  I pointed out incorrect assumptions on your part and issues with your suggestions.

I cannot speak for HaloFour, but seeing him in many issues, i can say that i do not get the sense that he "here to win the argument".  Indeed, i get the opposite sense.  That he (like myself), wants to really help peel back *all* the layers so that the understanding is complete.  It feels like you've interpreted this approach as being mean-spirited or "harsh", when the reality is (as @jnm2 mentioned) it's really just an attempt to provide you with **all** the data in a transparent fashion.  

In other words, when i see a person say "i think feature X should be like this because of M", i find that it's helpful to not just address "Y", but to also bring their attention to "N, O, P, Q, R and S", all of which were *essential* parts of the design that let to 'X'.  

I'll note that there was a lot of examples and explanations provided that you seemed to never actually address, giving hte impression that you were not paying attention to it or trying to understand it's place in the decision making.  That did not bother me, but indicated to me that it was important to help you out toward understanding why these parts were important.

I also noticed you tended to jump to conclusions a lot.  i.e you would look at one part of the language (like the 'where->Where' translation in linq) and assume that pattern was how we did thing everywhere (and this 'deconstruct' was out of place).  I often tried to show you that these things you thought were 'truisms' of the language, and the conclusions you were jumping to did not actually hold.

**Comment by CyrusNajmabadi** [comment]
> Although I have to reinforce that if you consider communication theory: Whoever is trying to provide information is responsible for the way it gets to the target.

This goes both ways :)

**Comment by CyrusNajmabadi** [comment]
> Well, those are just a few examples. 

Every one of those examples was actually addressed in the thread.  Often with many examples and deep explanations as to why it was felt that that perspective brought in more cons than pros.  :)

**Comment by CyrusNajmabadi** [comment]
> I'm just curious though: Besides flexibility (because we can transform data inside the Deconstruct method) why our deconstruction is not simply like Python's, ES6's or Elixir where you're a matching against the "public" members of the type automatically?

The primary reason is that there is no way to know how to property 'Deconstruct' an object.  It's entirely domain specific and is not baked into anythign in the type itself *by default*.  For example, say i have the following:

```c#
class WhileStatementSyntax {
    SyntaxToken WhileKeyword;
    SyntaxToken OpenParenToken;
    ExpressionSyntax Expression;
    SyntaxToken CloseParentToken;
    StatementSyntax Statement;
}
```

Now say i'm in a domain where i want to work with ASTs (i.e. **abstract** syntax trees).  What i'll actually want to do is:

```c#
var (expr, stmt) = whileStatement;
```

How can the language know that those 5 fields relate to those two deconstructed values i actually care about?  Similarly, let's say the language did have some sort of 'default' rule for how things could be deconstructed.  What would that rule be?   It's tempting to say:

"well.. it would be the readable fields in the order they're declared in the class".

But tha'ts pretty unpalatable.  For one thing, up till now, it has almost never mattered what order you put things in your types.  Many people have no order, or do it alphabetically, or do thins in the best order internally, but not with consideration to the common external way it would be deconstructed.  Also:

"Well.. what about properties".

Do you have both fields and properties?  Should they be grouped in any way?  Also:

"What about partial types?"

If i have a partial type with many members spread across teh parts... how do we come up with a deconstruction member that is sensible?  Also:

"What about metadata?"

Members are not ordered in metadata.  If the compiler was injecting default deconstructors for types, there would be no way it could know the right order for all the types you are referencing.

--

Basically, if we had had pattern-matching/deconstruction from *early* on, we would have likely baked something into the type itself, and then given prescriptive guidance on how people should shape their types.  But we're well past that point now.  There are billions of lines of code out there written over 15+ years.  People will rarely go back and rewrite their types to improve this situation.  

Because of that, we design features with a strong eye toward "how can we ensure this *will* get adopted in *todays* ecosystem".

--

Note: future work we're doing (with things like ADTs/records/primary-constructors/etc.) are much more "member order"-oriented. As such, those types likely *would* get automatic-deconstruction capabilities.  But that's because their design is explicitly thinking about this sort of thing, whereas the design of classes/structs/interfaces **20** years ago was not taking any of this into consideration.

**Comment by thiagomajesk** [comment]
@jnm2 

> Just want to point out that “if you care to” is an argumentative tone which drowns out the signals you wanted to communicate, and there are examples of this in your responses to the others.

This can be a poor choice of words, mainly because I'm a non-native speaker. So the only thing I can do about is to apologize if this came across the wrong way ;)

@CyrusNajmabadi

> I cannot speak for HaloFour, but seeing him in many issues, i can say that i do not get the sense that he "here to win the argument"

I was actually addressing that to you because that was my perception of your approach, not the way you have described you want it to be. Btw, I've read other Issues you have participated and this seems to be the general sense, so is not just me. Maybe you could reevaluate the way you are expecting to communicate your intentions.

> THat's why i spent a lot of time giving many different examples to try to help make things click for you. However, even when i responded, you lashed at at me. I never treated you in a disrespectful fashion

Maybe you didn't realize that, but like I pointed before, some of your comments really came across in a harsh fashion (like you've perceived some of mine) and maybe wasn't your intent either.
But just the fact that the information you provided was not "clicking" for me, is a good demonstration that perhaps you should have been more straight to the point like @HaloFour did while trying to address my understanding of the problem. (And this is why I pointed out I didn't think you had the proper "care" to address the issue).

> >Although I have to reinforce that if you consider communication theory: Whoever is trying to provide >> information is responsible for the way it gets to the target.
> 
> This goes both ways :)

Of course it does, but since you had the upper hand regarding the information of how all makes sense (for being in the design team), you are more responsible for making the context clear, while I'm only responsible for making my current understanding of the matter clear - and that blame is one me for the parts that apply. 
And you have complained about me jumping to conclusions, but how can I be to blame if I'm making my best effort to understand things you're are explaining partially a concept I'm not aware of!?
Is like explaining someone how gas fuels your special-car while expecting someone that doesn't know how you're special-engine works. For sure you can give a good explanation full of details that make sense for whoever is involved in the project, but it won't do much for the "curious" person who's asking.
And the responsibility is on the person with more information to know how to approach the other one.

> > Well, those are just a few examples.
> 
> Every one of those examples was actually addressed in the thread. Often with many examples and deep explanations as to why it was felt that that perspective brought in more cons than pros. :)

Even though you say that you have addressed those questions, just until @HaloFour jumped into the thread to explain objectively what and how I was getting the wrong concept that things were made clear. So, can we agree that maybe you could have tried a different approach?

> In such a discussion it is very difficult to communicate because i only have your words to gauge your understanding. In a face to face conversation, i would be able to pick up on an enormous number of clues as to what was going on. But here, i can only use what you say. And so, when you are silent and do not actually address the feedback and questions about your points, it makes the conversation much more difficult and ambiguous.

Yes, I agree with that. But I can escalate that for both parts and ambiguity will still happen if you do not try to address the problem from the ground up and instead wait for feedback on something I may not know how to respond to. Again, like @HaloFour did with a lot less grasp of the conversation than you; he has picked those clues and addresses the problem very quickly.

>Consider how confusing that can make things and how that puts hte explainer in an even more difficult position as i now need to consider all of those above '4' cases across many parts of the conversation when discussing things with you.

I fully recognize that, but look at how much you have written to address my comments. By doing that you extrapolated the whole conversation to address multiple points instead of addressing a specific one, clearing that out and moving to another. 
Of course in a "real world conversation" you would have more clues about the conversation, but unfortunately this is the constraint we are in, so we are even more responsible for the outcoming of our interactions - and of course this is both ways, but like I said before: _you have the upper-hand regarding the right context and information_.

If I can illustrate my perception of your approach again: Is like a teacher blaming a student for not knowing how to formulate a question while trying to grasp a complex concept. What you normally expect from a reasonable approach is that the person who is trying to deconstruct the wrong idea will lead the way, not the other way around. The difference is if I would explain something to you, I'd not inflict you the responsibility to provide the other part of the conversation unless I know 100% for sure you are understating how things work. Otherwise, we'll for sure, have a lot of loose ends in a conversation, because you are expecting someone to provide information they don't have as you push forward.

Although is true (like someone already mentioned before) that is hard to understand how much someone knows about the topic through text conversations, I think this is a matter of facing the explanation in a more informative-way than a counter-argumentative one. 
I'm basing this rationale around the fact that you should care only to explain the motivation behind the decisions instead of counterpointing everything like I saw HaloFour doing effectively.

@CyrusNajmabadi I hope I could better point out the deficiencies I felt on the way you have approached me with this subject so you can improve on that. Since the language is expanding beyond the niche it was a couple of years ago (since opensource), you guys at Microsoft will have to deal with people from many different backgrounds and understandings with a range of different experiences and views. And while I know communication is not a strong skill set for many in our field, is really important for dealing with the community (And of course these words are bidirectional).

Well, since we have diverged a lot from the initial point of the discussion, I don't think it will be any more productive to keep on discussing this. Since all points were made clear for both parts regarding what had to be spoken and now we have accumulated a good history of the conversation, maybe this thread should be put to rest. Unless of course, that someone else has some relevant information to bring in to the topic.

**Comment by CyrusNajmabadi** [comment]
>  By doing that you extrapolated the whole conversation to address multiple points instead of addressing a specific one, clearing that out and moving to another.

You brought up many points.  And, importantly, a large part of the reasons i think the current C# approach is good because *in aggregate* the pros it brings across all those points outweighs negatives you see.  I felt if i only focused on one area that would do you a disservice because in any one area you *could* potentially argue the value just wasn't present.  I also did not want to get bogged down into only one area, making it seem as if this was the only reason for the decision making.

I presented all parts of the decision to make the reasoning as transparent as possible.  I find that this approach works well for many people (myself included) so i started with that as the default.  Again, i'm not a mind reader.  I don't know what will work best for you, and especially give your *lack* of engagement and responses on many points brought up, i had no reason to believe that focusing on one area would be any more successful.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi I hope I could better point out the deficiencies I felt on the way you have approached me with this subject so you can improve on that.

> Well, since we have diverged a lot from the initial point of the discussion, I don't think it will be any more productive to keep on discussing this. 

I'm very active over on gitter.im. Feel free to come chat in https://gitter.im/dotnet/csharplang or https://gitter.im/dotnet/roslyn or just direct-message me.  We can chat further if you feel there are ways i could provided my examples and information in a way thta would have made more sense for you and for future readers.

**Comment by thiagomajesk** [comment]
@CyrusNajmabadi I'll be glad to do that in the future if necessary, thank you for being open-minded about my final considerations 😉 👍.

**Comment by CyrusNajmabadi** [comment]
Definitely.  Thanks for the feedback!

**Comment by jnm2** [comment]
> > Just want to point out that “if you care to” is an argumentative tone which drowns out the signals you wanted to communicate, and there are examples of this in your responses to the others.
> 
> This can be a poor choice of words, mainly because I'm a non-native speaker. So the only thing I can do about is to apologize if this came across the wrong way ;)

If that's a factor for you, the other thing you can do is give folks like Cyrus the benefit of the doubt when they don't understand where you're coming from right away.

**Comment by YairHalberstadt** [comment]
Closing as the addition of an interface is a request for dotnet/runtime.

If the proposal is to restrict deconstruction to the interface that would be a breaking change, and is not viable.
